pkg/comp-functions/functions/vshnnextcloud/collabora.go

```diff
@@ -0,0 +1,597 @@
+package vshnnextcloud
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+
+	"crypto/x509"
+	_ "embed"
+	"encoding/pem"
+	"fmt"
+
+	valid "github.com/asaskevich/govalidator"
+	cmv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	certmgrv1 "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
+	xfnproto "github.com/crossplane/function-sdk-go/proto/v1beta1"
+	vshnv1 "github.com/vshn/appcat/v4/apis/vshn/v1"
+
+	"github.com/vshn/appcat/v4/pkg/comp-functions/functions/common"
+	"github.com/vshn/appcat/v4/pkg/comp-functions/runtime"
+	"gopkg.in/yaml.v2"
+	v1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/ptr"
+)
+
+var (
+	watchOnlySecretObjectName = "-collabora-code-wo-secret"
+	ingressCertificateName    = "-collabora-code-ingress-certificate"
+	labelMap                  = map[string]string{runtime.WebhookAllowDeletionLabel: "true"}
+)
+
+//go:embed files/coolwsd.xml
+var coolwsd string
+
+//go:embed files/coolkit.xml
+var coolkit string
+
+func DeployCollabora(ctx context.Context, comp *vshnv1.VSHNNextcloud, svc *runtime.ServiceRuntime) *xfnproto.Result {
+
+	err := svc.GetObservedComposite(comp)
+	if err != nil {
+		return runtime.NewWarningResult("Cannot GetObservedComposite: " + err.Error())
+	}
+
+	if !comp.Spec.Parameters.Service.Collabora.Enabled {
+		return runtime.NewNormalResult("Collabora not enabled")
+	}
+
+	if !valid.IsDNSName(comp.Spec.Parameters.Service.Collabora.FQDN) {
+		return runtime.NewWarningResult("Collabora FQDN is not a valid DNS name: " + comp.Spec.Parameters.Service.Collabora.FQDN)
+	}
+
+	rules := []rbacv1.PolicyRule{
+		{
+			APIGroups: []string{""},
+			Resources: []string{"pods", "pods/log", "pods/status", "pods/attach", "pods/exec"},
+			Verbs:     []string{"get", "list", "watch", "create", "watch", "patch", "update", "delete"},
+		},
+		{
+			APIGroups: []string{"apps"},
+			Resources: []string{"deployments", "statefulsets"},
+			Verbs:     []string{"get", "delete", "watch", "list", "patch", "update", "create"},
+		},
+	}
+
+	if svc.Config.Data["isOpenshift"] == "true" {
+		rules = append(rules, rbacv1.PolicyRule{
+			APIGroups:     []string{"security.openshift.io"},
+			Resources:     []string{"securitycontextconstraints"},
+			ResourceNames: []string{"appcat-collabora"},
+			Verbs:         []string{"use"},
+		})
+	}
+
+	err = common.AddSaWithRole(ctx, svc, rules, comp.GetName(), comp.GetInstanceNamespace(), "collabora-code", true)
+	if err != nil {
+		return runtime.NewWarningResult("Failed to add Collabora SA with Role: " + err.Error())
+	}
+
+	// Create coolwsd config map
```

Comment:
why bother with such comments? you have exactly the sentence in the name of the function